src/plugins/vis_builder/public/application/components/data_tab/index.tsx

```diff
@@ -3,19 +3,336 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import React from 'react';
+import React, { useEffect, useState } from 'react';
+import { DropResult, EuiDragDropContext } from '@elastic/eui';
 import { FieldSelector } from './field_selector';
 
 import './index.scss';
 import { ConfigPanel } from './config_panel';
+import { useAggs, useVisualizationType } from '../../utils/use';
+import { useTypedDispatch, useTypedSelector } from '../../utils/state_management';
+
+import {
+  reorderAgg,
+  updateAggConfigParams,
+} from '../../utils/state_management/visualization_slice';
+import { BucketAggType, IndexPatternField, propFilter } from '../../../../../data/common';
+import { useOpenSearchDashboards } from '../../../../../opensearch_dashboards_react/public';
+import { VisBuilderServices } from '../../../types';
 
 export const DATA_TAB_ID = 'data_tab';
 
+const getIndexPatternField = (indexFieldName: string, availableFields: IndexPatternField[]) =>
+  availableFields.find(({ name }) => name === indexFieldName);
+
+const filterByName = propFilter('name');
+const filterByType = propFilter('type');
+
+export const getValidAggTypes = ({
+  fieldName,
+  sourceGroup,
+  destinationSchema,
+  aggProps,
+  aggService,
+  sourceAgg,
+}) => {
+  const isCountField =
+    sourceGroup === 'preDefinedCountMetric' ||
+    (sourceAgg && Object.keys(sourceAgg.params).length === 0);
+
+  const indexField = isCountField
+    ? { type: 'count' }
+    : getIndexPatternField(fieldName, aggProps.indexPattern?.fields ?? []);
+
+  if (!indexField) return [];
+
+  // Get all aggTypes allowed by the schema and get a list of all the aggTypes that the dragged index field can use
+  const aggTypes = aggService.types.getAll();
+  // `types` can be either a Bucket or Metric aggType, but both types have the name property.
+  const allowedAggTypes = filterByName(
+    aggTypes[destinationSchema.group] as BucketAggType[],
+    destinationSchema.aggFilter
+  );
+
+  return (
+    allowedAggTypes
+      .filter((aggType) => {
+        const allowedFieldTypes = aggType.paramByName('field')?.filterFieldTypes;
+        return filterByType([indexField], allowedFieldTypes).length !== 0;
+      })
+      .filter((aggType) => (isCountField ? true : aggType.name !== 'count'))
+      // `types` can be either a Bucket or Metric aggType, but both types have the name property.
+      .map((aggType) => (aggType as BucketAggType).name)
```

Comment:
I don't quite get this part. If types can be both metric or bucket agg, why we assign only bucket agg type here `aggType as BucketAggType`?